docs/proposals/kubelet-qos.md

```diff
@@ -0,0 +1,142 @@
+# Quality of Service in Kubernetes
+
+**Author**: Ananya Kumar (@AnanyaKumar)
+
+**Status**: Draft proposal; prototype in progress.
+
+*This document presents the design of quality of service tiers for pods in Kubernetes, describes use cases, and lays out milestones for its development.*
+
+## Motivation
+
+Kubernetes allocates resources to pods in a simple way. Users specify limits for containers. For example, a container can have a 1gb memory limit. The scheduler uses the limit to launch containers. For example, the scheduler will not place 5 containers with a 1gb memory limit into a machine with 4gb memory. Additionally, Kubelet does not allow containers to exceed specified limits. In the current implementation, there aren’t robust mechanisms in place to ensure that pods run reliably on an overloaded system and nodes are not very efficiently utilized.
+
+**Pods with limits specified are not very well protected from pods without limits specified**. If a pod without an explicit memory limit goes overboard and uses lots of memory, other pods (with specified memory limits) might be killed. This is bad, because users often want a way to launch containers that have resources guarantees, and that stay up reliably.
+
+On the other hand, **if all pods have specified limits, cluster utilization can be poor**. In the current implementation, Kubelet does not support over-committing pods. Pods often don’t use all the resources that they request which leads to a lot of wasted resources. For example, we might have 4 pods, each reserving 1GB of memory in a node with 4GB memory, but only using 500MB of memory. Theoretically, we could slot in more pods onto the node, but Kubernetes will not schedule new pods on the node.
+
+This proposal provides mechanisms for solving both the reliability and efficiency problems, by allowing pods to specify levels of resource guarantees. Containers will be able to *request* for a minimum resource guarantee. The *request* is different from the *limit* - containers will not be allowed to exceed resource limits.
+
+Additionally, we will support 3 different tiers of pods. Guaranteed pods (where requests and limits are equal) will receive the highest priority and will get all the resources they need. Burstable pods, which specify request < limit will be guaranteed their request and can “burst” and use more resources when available. Best-Effort pods, which don’t specify a request, can use resources only if not being used by other pods. In Borg, supporting best-effort pods increased utilization by about 20%, and we hope to see similar improvements in Kubernetes.

```

Comment:
Hmm. This definition has a few problems:
* The best-effort QoS tier won't be usable until we can perform worst-fit scheduling based on derived usage metrics. We do plan to do that, but it's not done yet.
* We currently use LimitRange to reserve minimum resources for containers, which causes them to be spread around using the current scheduling policy. I proposed changes to LimitRange here: https://github.com/GoogleCloudPlatform/kubernetes/pull/12035#issuecomment-126803371. However, even with those changes, it would not be able to bound resource usage (min/max) while still allowing the best-effort QoS.
* In the absence of LimitRange, it doesn't mitigate gaming, since a very large Limit and very small Request could be set, putting the container resource into the Burstable class.

I think we can go ahead, but probably will want to refine the definition. Maybe ratio of limit to request. Limit 100x Request isn't really Burstable. Probably at most 5x or 10x would make sense. The admin might want to configure it. LimitRange would be a logical place for that. MaxBurstableLimitToRequestRatio (Or maybe MaxBurstableRatio for short). 

@derekwaynecarr WDYT?